### DISCUSSION TITLE
Unified switch-when, if, catch-when

### DISCUSSION THREAD

**Discussion by zms9110750** [dotnet/csharplang#8396] [General]
The usage of 'when' and 'if' on switch is similar. But their keywords are different and there are no parentheses.
Their functions are the same, but they use completely different syntax.

Moreover, due to the support of pattern matching, multiple catches can be completely replaced by switches.
So now, some of these three functions are unnecessary.

And the swtich expression can discard+when, but the switch option cannot use default+when

**Comment by zms9110750** [comment]
I know that grammar that only changes the style without actually solving the problem will not be supported.
But I still want to write it down.

```csharp
try
{
}
catch switch
{
	case IOException e if （e.Message.Length < i）:
		break;
	case if (i < 3):
	default:
		break;
}
```

**Comment by HaloFour** [comment]
> The usage of 'when' and 'if' on switch is similar. But their keywords are different and there are no parentheses.

This is intentional as they are guard clauses, not conditional statements.  So the language uses `when` just like it does for guard clauses with exception handling.

> And the swtich expression can discard+when, but the switch option cannot use default+when

You can use discard + when in switch statements:

```cs
switch (...) {
    case var _ when condition:
        break;
}
```

As for mixing catch with switch, check out:

#989
